### PR TITLE
Add "field ID's" to new serialization

### PR DESCRIPTION
**Pull request by Maxxen**
This PR adds "field-ID's" to the new serialization infrastructure. Inspired by protobuf and other [TLV](https://en.wikipedia.org/wiki/Type–length–value) encoding schemes, these will hopefully aide in making the binary serialization format backwards (and maybe even future) compatible. Right now we just write and verify that we get the expected ones back when deserializing, but to actually get all the backwards/future compatability goodness I'm going to have to make some changes to the format itself and how the serializer works. I'll do that in a follow-up PR, which will also include tests that showcase how these field ids help with compatibility between versions. 

But since just these changes touches a lot of files I figured its best to get them merged ASAP.

### CODE REVIEW

**Comment by Maxxen** [comment]
Got a failing dev build due to variable used in an assert being unused in release mode, pushing again.

**Comment by carlopi** [comment]
Failures (Java job + potentially Substrait failure) looks to be unrelated to this PR.

File: src/include/duckdb/common/serializer/binary_deserializer.hpp [hunk 1]
```diff
@@ -45,13 +45,17 @@ class BinaryDeserializer : public FormatDeserializer {
 	struct State {
 		uint32_t expected_field_count;
 		idx_t expected_size;
+		field_id_t expected_field_id;
```
Comment (taniabogatsch) [inline src/include/duckdb/common/serializer/binary_deserializer.hpp:48]: Does `expected_field_count` mean that there are multiple fields? Or is this the size of one field (`expected_field_size` might be easier to understand)? And if multiple fields exist, why is there only one `expected_field_id`?

I am leaving these questions here to document my code-reading process a bit...

File: src/include/duckdb/common/serializer/binary_deserializer.hpp [hunk 2]
```diff
@@ -72,7 +76,7 @@ class BinaryDeserializer : public FormatDeserializer {
 	}
 
 	// Set the 'tag' of the property to read
-	void SetTag(const char *tag) final;
+	void SetTag(field_id_t field_id, const char *tag) final;
```
Comment (taniabogatsch) [inline src/include/duckdb/common/serializer/binary_deserializer.hpp:79]: Nit: `const`?

File: src/include/duckdb/common/serializer/binary_deserializer.hpp [hunk 3]
```diff
@@ -45,13 +45,17 @@ class BinaryDeserializer : public FormatDeserializer {
...
 		uint32_t expected_field_count;
 		idx_t expected_size;
+		field_id_t expected_field_id;
 		uint32_t read_field_count;
-		State(uint32_t expected_field_count, idx_t expected_size)
-		    : expected_field_count(expected_field_count), expected_size(expected_size), read_field_count(0) {
+
+		State(uint32_t expected_field_count, idx_t expected_size, field_id_t expected_field_id)
+		    : expected_field_count(expected_field_count), expected_size(expected_size),
+		      expected_field_id(expected_field_id), read_field_count(0) {
```
Comment (taniabogatsch) [inline src/include/duckdb/common/serializer/binary_deserializer.hpp:53]: Also, it might be interesting to use `const` here to clarify which are mutable and which aren't... Sorry for all these `const` comments... 😅

File: src/include/duckdb/common/serializer/binary_serializer.hpp [hunk 4]
```diff
@@ -55,7 +56,7 @@ struct BinarySerializer : public FormatSerializer {
 		return std::move(serializer.data);
 	}
 
-	void SetTag(const char *tag) final;
+	void SetTag(field_id_t field_id, const char *tag) final;
```
Comment (taniabogatsch) [inline src/include/duckdb/common/serializer/binary_serializer.hpp:59]: Nit: `const`?

File: src/include/duckdb/common/serializer/format_deserializer.hpp [hunk 5]
```diff
@@ -392,7 +392,8 @@ class FormatDeserializer {
 	}
 
 protected:
-	virtual void SetTag(const char *tag) {
+	virtual void SetTag(field_id_t field_id, const char *tag) {
+		(void)field_id;
```
Comment (taniabogatsch) [inline src/include/duckdb/common/serializer/format_deserializer.hpp:396]: Just copy-pasting my previous comment here...
> Nit: `const`? And do we need `(void)field_id`, as in, does some tidy check complain here if we do not use the parameter? I don't think it does, e.g., for function implementations in duckdb... 🤔

File: src/include/duckdb/common/serializer/format_deserializer.hpp [hunk 6]
```diff
@@ -30,22 +30,22 @@ class FormatDeserializer {
 public:
 	// Read into an existing value
 	template <typename T>
-	inline void ReadProperty(const char *tag, T &ret) {
-		SetTag(tag);
+	inline void ReadProperty(field_id_t field_id, const char *tag, T &ret) {
```
Comment (taniabogatsch) [inline src/include/duckdb/common/serializer/format_deserializer.hpp:33]: I feel a bit weird by now, another `const` comment... 😅 But would it make sense to add `const` here to all the function parameters?

File: src/include/duckdb/common/serializer/format_serializer.hpp [hunk 7]
```diff
@@ -196,7 +196,8 @@ class FormatSerializer {
 	}
 
 	// Handle setting a "tag" (optional)
-	virtual void SetTag(const char *tag) {
+	virtual void SetTag(field_id_t field_id, const char *tag) {
+		(void)field_id;
```
Comment (taniabogatsch) [inline src/include/duckdb/common/serializer/format_serializer.hpp:200]: See `format_deserializer.hpp` for this and the other function definitions.

File: src/include/duckdb/common/serializer/serialization_traits.hpp [hunk 8]
```diff
@@ -10,6 +12,8 @@ namespace duckdb {
 class FormatSerializer;   // Forward declare
 class FormatDeserializer; // Forward declare
 
+typedef uint16_t field_id_t;
```
Comment (taniabogatsch) [inline src/include/duckdb/common/serializer/serialization_traits.hpp:15]: Possibly bump this to `uint32_t`, just to have slightly more flexibility.

**Review by taniabogatsch — CHANGES_REQUESTED** [review]
Hey @Maxxen, thanks for the PR! My comments are mostly nitpicks and questions. I also have some more general points. 😄 

- I think we should add some developer notes once this becomes our default way for (de)serialization. These notes should contain what to remember when altering classes or adding/deleting them.
- I maybe missed it, but do we have a test that checks that there are no accidental duplicate field IDs in the `.json` files?

**Review by carlopi — APPROVED** [review]
As mentioned to @Maxxen, work looks cool to me.

One potential improvement, to be considered if/when this makes sense, is having the generate_serialization.py script have an optional behaviour of generating ID's for fields missing them.

Current behaviour (introduced in this PR) checks whether ID's make sense, but potentially the same logic could also be used to generate ids on the fly: say you get a .json missing some id fields, and they are generated to some valid value and added to the .json files.

Basically you move it to 2 steps:
1. check IDs (and potentially adds them if needed). (json in input are checked, and potentially fields added to them)
2. generate (de)serialization c++ code (json as input -> cpp/hpp as output)

But once again, this looks cool, thanks.

File: extension/json/json_deserializer.cpp [hunk 9]
```diff
@@ -3,7 +3,7 @@
 
 namespace duckdb {
 
-void JsonDeserializer::SetTag(const char *tag) {
+void JsonDeserializer::SetTag(uint16_t field_id, const char *tag) {
```
[the remark this answers is not in the document]
Comment (Maxxen) [reply extension/json/json_deserializer.cpp:6]: Nope!

File: extension/json/json_deserializer.cpp [hunk 10]
```diff
@@ -122,11 +122,11 @@ void JsonDeserializer::OnMapEntryBegin() {
 }
 
 void JsonDeserializer::OnMapKeyBegin() {
-	SetTag("key");
+	SetTag(100, "key");
```
[the remark this answers is not in the document]
Comment (Maxxen) [reply extension/json/json_deserializer.cpp:125]: For the JSON case it doesn't matter since it doesn't use field Ids anyway, but we need to set something...

File: src/include/duckdb/common/serializer/format_deserializer.hpp (hunk 5 above)
Comment (Maxxen) [reply src/include/duckdb/common/serializer/format_deserializer.hpp:396]: Yeah I'm getting "unused parameter" warnings (which turn to failures) otherwise so I (void) it here.

File: src/include/duckdb/common/serializer/serialization_traits.hpp (hunk 8 above)
Comment (Maxxen) [reply src/include/duckdb/common/serializer/serialization_traits.hpp:15]: Agree, done!

File: extension/json/include/json.json [hunk 11]
```diff
@@ -6,18 +6,22 @@
     ],
     "members": [
       {
+        "id": 100,
```
[the remark this answers is not in the document]
Comment (Maxxen) [reply extension/json/include/json.json:9]: For now, yes. Note that they don't actually have to be incremental, and when generating the code from the JSON we check that you don't accidentally overlap the ID with any parent or child classes. Although we may provide some way for the script to auto-generate and save these for "new" untagged field.

**Comment by Maxxen** [comment]
Thanks for the thorough review @taniabogatsch! to answer you questions

> I think we should add some developer notes once this becomes our default way for (de)serialization. These notes should contain what to remember when altering classes or adding/deleting them.

I agree, I'm going to write up some more docs on this once I make more changes to the binary encoding itself.

> I maybe missed it, but do we have a test that checks that there are no accidental duplicate field IDs in the .json files?

Not a test per se, but the generator script will traverse the class hierarchy and exit with an error message if there is a duplicate field id or field name in any inheritance chain. In the future I will add additional checks to the FormatSerializer at runtime when running in debug mode to ensure it doesn't see any duplicate ids. These checks will then run in all of our test suites where we run with verification and round-trip every statement.

File: extension/json/json_deserializer.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (taniabogatsch) [reply extension/json/json_deserializer.cpp:125]: Hmm, I am wondering if this is confusing. Would it make sense to reverse the order of parameters to allow for default parameters? But might be something to consider in a different PR.

File: src/include/duckdb/common/serializer/format_deserializer.hpp (hunk 5 above)
Comment (taniabogatsch) [reply src/include/duckdb/common/serializer/format_deserializer.hpp]: Could you solve this the same way as I suggested in my other comment? `SetTag(const char *tag, field_id_t field_id = 0)`, something like that? I am wondering what an `INVALID_FIELD_ID` would look like, though, since this is `uint32_t`...
> Hmm, I am wondering if this is confusing. Would it make sense to reverse the order of parameters to allow for default parameters? But might be something to consider in a different PR.

Comment (Maxxen) [reply src/include/duckdb/common/serializer/format_deserializer.hpp]: Since field_ids are required I don't want to set a default parameter and have people accidentally leave it out. In theory I could probably replace it with `SetTag(const char *tag, const field_id_t)` but it doesn't matter for the compiler.

Comment (Maxxen) [reply src/include/duckdb/common/serializer/format_deserializer.hpp]: Would it be alright if I tweaked that in a follow up PR? Since this already covers a lot of surface area and Im kind of under time pressure I really want to get this merged ASAP so I can continue with the next steps.

**Review by taniabogatsch — APPROVED** [review]

**Comment by carlopi** [comment]
One thing I would consider, at some later point, is separating validation of json files from generation of C++ files.

This came to my mind connected to https://github.com/duckdb/duckdb/pull/8536#pullrequestreview-1573683251, but might stand even independently.
Basically a first step would check / validate json files (and eventually raise an error and terminate) while a second step assumes json to be valid, and only touches the cpp files (without checking for errors).

Those two step can partially reuse the same underlying logic, and can also be implemented as part of the same script using parameters, but will make so that state is never inconsistent (like generation failing mid-way due to validation error).

**Comment by hannes** [comment]
Thanks everyone!
